internal/item/item.go

```diff
@@ -0,0 +1,269 @@
+package item
+
+import (
+	"errors"
+	"time"
+
+	"github.com/ardanlabs/kit/db"
+	"github.com/ardanlabs/kit/db/mongo"
+	"github.com/ardanlabs/kit/log"
+
+	gc "github.com/patrickmn/go-cache"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// Contains the name of Mongo collections.
+const (
+	Collection        = "coral_items"
+	CollectionHistory = "coral_items_history"
+)
+
+// Set of error variables.
+var (
+	ErrNotFound = errors.New("Item Not found")
+)
+
+// =============================================================================
+
+// c contans a cache of set values. The cache will maintain items for one
+// second and then marked as expired. This is a very small cache so the
+// gc time will be every hour.
+
+const (
+	expiration = time.Second
+	cleanup    = time.Hour
+)
+
+var cache = gc.New(expiration, cleanup)
+
+// =============================================================================
+
+// create an item out of its type, version and data or die trying
+//  if the item already exists, adopt it's Id but do not presume to update
+func Create(context interface{}, db *db.DB, t string, v int, d ItemData) (Item, error) {
+
+	i := Item{}
+
+	// Check to see if this item already exists in the db
+	//  this is done by checking it's "foreign identity":
+	//  Checking for its type and it's Type.Id fields
+
+	// get this data's IdField value for this type
+	idValue := getDatumByKey(Types[t].IdField, d)
+
+	// if there is not a value, generate a new id and continue
```

Comment:
From effective go: 

```
It's therefore possible—and idiomatic—to write an if-else-if-else chain as a switch.
```

So, you could switch on `idValue`.  Although there isn't really a long chain of if, else here.